src/solvers/slvr_blk_2m.hpp

```diff
@@ -100,8 +101,8 @@ class slvr_blk_2m<
         for (int j = this->j.first(); j <= this->j.last(); ++j)
         {
           auto
-          dot_rr = rhs.at(parent_t::ix::rr)(i, j, this->k),
-          dot_nr = rhs.at(parent_t::ix::nr)(i, j, this->k);
+          dot_rr = this->rr_flux(i, this->j),
```

Comment:
@claresinger The only obvious error I found is the one above, but it only affects 3D simulations.

I checked again bulk microphysics in libcloudph++ and realized (again) that rhs_columnwise computes the change in rr/nr due to sedimentation, which is equal to the inflow from the cell above minus the outflow to the cell below. That's why you get negative values. In blk_1m it's the same. I suppose that the"precipitation_rate" name is misleading.

If you want to plot outflow, you need to substract inflow from the cell above (starting from the top of the column). This can be done in post-processing, i.e. at the moment of plotting.
In the UWLCM_plotting package:
https://github.com/igfuw/UWLCM_plotting
there is a UWLCM_plotters class that has a "load_prflux_timestep" function that does this (right now for blk_1m only, but for blk_2m it would be almost the same).
You could use this class directly, the "drawbicyc" program that uses UWLCM_plotters and can plot slices like the one you've shown, or just adapt your own plotting scripts.

Reply:
@pdziekan This seems very incorrect - too spiky, and also, the magnitude seems low (by a factor of 100 maybe), compared to the intercomparison paper (and the `lgrngn` simulations). I _think_ my plotting script is doing the same thing as here. https://github.com/igfuw/UWLCM_plotting/blob/8495f53e526389687946afa3ec2afa24b000b1ba/UWLCM_plotters/include/PlotterMicro.hpp#L114-L123

Are you sure that the 1M code I have been using as a template is correct? Do you have a plot of the precipitation flux from the 1M? 

![image](https://github.com/igfuw/UWLCM/assets/6875824/acb10b6f-6445-4835-b90f-9e346be58960)
![output](https://github.com/igfuw/UWLCM/assets/6875824/5aabbbf0-9749-4292-9bf3-4f7ad2c09a42)

```
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr

path = "../../output_full/dycoms_blk2m/dycoms_default_precip/"
const = xr.open_dataset(path+"exp1/const.h5")
rhod = const.rhod.values # kg/m3
Lv = 2.5e6 # J/kg

f = path+"exp1/timestep0000006000.h5"
ds = xr.open_dataset(f)
dz = 5 # m 
dx = 50 # m
x = dx/2 + ds.phony_dim_0*dx # m
z = dz/2 + ds.phony_dim_2*dz # m
time = np.array(float(f.split("/")[-1].split(".")[0].split("timestep")[-1])) # s

precip_flux = np.copy(ds.precip_rate_rr.values) # kg/kg
for i in np.arange(len(z)-1):
    k = len(z) - i - 2
    precip_flux[:, k] = precip_flux[:, k+1] - precip_flux[:, k]
precip_flux *= dz * Lv * rhod

fig,axes = plt.subplots(1,2,sharey=True,figsize=(10,5),gridspec_kw={"width_ratios":[0.3,0.7]})
ax = axes[0]
ax.plot(np.mean(precip_flux, axis=0), z, color="darkred")
ax.set_ylabel("z [m]")
ax.set_xlabel("Precip flux [W/m2]")

ax = axes[1]
p = ax.contourf(x, z, precip_flux.T,levels=np.linspace(0,0.01,11),cmap="Reds",extend="max")
plt.colorbar(p, label="Precip flux [W/m2]")
plt.xlabel("x [m]")

plt.suptitle("t = {:.0f} s".format(time), y=0.95)
plt.show()
```

Reply:
Maybe there's no cloud at this moment?

Reply:
![image](https://github.com/igfuw/UWLCM/assets/6875824/9a94e446-486c-4e56-892f-405b9b372409)
^^ added contours of cloud water content

Reply:
Ok, it does look wrong :) I'm looking at blk_1m. It seems that there are NaNs in Release mode, but not in Debug mode.

Reply:
I have fixed blk_1m simulations, new igfuw/master should work correctly.
I tested your python script on blk_1m results and the precipitation flux looks fine:
![Figure_1](https://github.com/igfuw/UWLCM/assets/10848601/246785d9-35b9-479d-9a9f-8b9af177fa91)

Reply:
Thanks @pdziekan! I will check the blk2m results now!

Reply:
Thanks @pdziekan - `blk1m` looks good (I can reproduce the above result locally). But the `blk2m` still has this very spiky, irregular appearance, and the magnitude is way off.
![image](https://github.com/igfuw/UWLCM/assets/6875824/bd9a9f9d-8032-4112-bd57-b1ef6d2e7e0f)

Reply:
Just to confirm, the same problems exist after a 1 hr spinup + 1 hr simulation. 
![image](https://github.com/igfuw/UWLCM/assets/6875824/83228ca6-4aef-4b36-a31f-7d742a6c55d9)

From comparing the implementation of blk2m with blk1m in UWLCM, I would guess that there either is a bug in `slvr_blk_2m_common.hpp` or in `libcloudph++` itself.

Reply:
#160 fixed some errors in blk_2m, but precip is still way too small.
I will look into it tomorrow.